README.md
Job's Done
===========

[![link](https://img.shields.io/pypi/v/jobs_done10.svg)](https://pypi.org/project/jobs_done10)
[![link](https://img.shields.io/pypi/pyversions/jobs_done10.svg)](https://pypi.org/project/jobs_done10)
[![tests](https://github.com/ESSS/jobs_done10/actions/workflows/tests.yml/badge.svg)](https://github.com/ESSS/jobs_done10/actions/workflows/tests.yml)
[![link](https://img.shields.io/github/license/ESSS/jobs_done10.svg)](https://img.shields.io/github/license/ESSS/jobs_done10.svg)
# About #

Job's Done is a tool heavily inspired by [Travis](https://travis-ci.org/), and works in the same way
in that configuring a `.jobs_done.yaml` file in your repository's root to create and trigger Continuous Integration jobs.

Example of a `.jobs_done.yaml` file:

```yaml
matrix:
  platform:
  - "win64"
  - "redhat64"

platform-win64:build_batch_commands:
- |
  python -m venv .env3   || goto eof
  call .env3\Scripts\activate   || goto eof
  pytest --junitxml=tests-{platform}.xml

platform-redhat64:build_shell_commands:
- |
  python3 -m venv .env3
  source .env3/bin/activate
  pytest --junitxml=tests-{platform}.xml

junit_patterns:
- "tests.*.xml"
```

Considering this file is in the root of repository `myproject` and was pushed to branch `feat-71`, this will generate two Jenkins jobs:

* `myproject-feat-71-win64`
* `myproject-feat-71-linux64`


## Command-line ###

Jobs done can be executed in the command-line.

To use it, from the repository's folder that you want to create jobs for, execute:

```console
$ jobs_done jenkins --username USER https://example.com/jenkins
```

This will create/update existing jobs.

Below are the possible installation options.

### PyPI ###

1. Create a virtual environment using Python 3.10 and activate it:

   ```console
   $ python -m venv .env
   $ .env\Scripts\activate  # Windows
   $ source .env/bin/activate  # Linux
   ```

3. Install jobs_done10:

   ```console
   $ pip install jobs_done10
   ```

### Development ###

1. Clone the repository:

   ```console
   git clone git@github.com:ESSS/jobs_done10.git
   cd jobs_done10
   ```

2. Create a virtual environment using Python 3.10 and activate it:

   ```console
   $ python -m venv .env
   $ .env\Scripts\activate  # Windows
   $ source .env/bin/activate  # Linux
   ```

3. Install dependencies and the project in editable mode:

   ```console
   $ pip install -r requirements.txt -e .
   ```


#### Upgrading dependencies

We use [pip-tools](https://pypi.org/project/pip-tools) to pin versions, follow the instructions in the
docs to add new libraries or update existing versions, adding `--extra dev` to include development dependencies.


## Server ##

jobs done includes a `flask` server in `jobs_done10.server` which can be deployed using [Docker](https://www.docker.com/).

It provides two end-points for `PUSH` and `GET`:

* `/stash`: tailored to receive the push event from BitBucket Server (formerly known as Stash).
* `/github`: tailored to receive the push event from GitHub. **Important**: the webhook must be configured with content-type of `application/json` and a **secret**.

A `GET` to either ends points will return the JobsDone version, useful to check the installed version and
that the end-point is correct.

Posting to `/` is the same as posting to `/stash` for backwards compatibility (might be removed in the future).

### Configuration ###

Configuration is done by having a `.env` file (cortesy of [python-dotenv](https://github.com/theskumar/python-dotenv))
in the root of this repository with the following variables:

```ini
JD_JENKINS_URL=https://example.com/jenkins
JD_JENKINS_USERNAME=jenkins-user
JD_JENKINS_PASSWORD=some password

JD_STASH_URL=https://example.com/stash
JD_STASH_USERNAME=stash-user
JD_STASH_PASSWORD=some password

JD_GH_TOKEN=github-user-personal-access-token
JD_GH_WEBHOOK_SECRET=webhook-secret

JD_EMAIL_USER=mail-sender@example.com
JD_EMAIL_FROM=JobsDone Bot <mail-sender@example.com>
JD_EMAIL_PASSWORD=email password
JD_EMAIL_SERVER=smtp.example.com
JD_EMAIL_PORT=587
```

### Build ###

Clone the repository and checkout the tag:

```console
$ git clone https://github.com/ESSS/jobs_done10.git
$ cd jobs_done10
$ git checkout <VERSION>
```

Build a docker image:

```console
$ docker build . --tag jobsdone:<VERSION> --build-arg SETUPTOOLS_SCM_PRETEND_VERSION=<VERSION>
```

### Run server ###

```console
$ docker run --publish 5000:5000 jobsdone:<VERSION>
```

# Hello World #

This is an example of a Job's Done file, and what you might expect of its contents.

```yaml
build_batch_commands:
- "echo MESSAGE: Hello, world!"

description_regex: "MESSAGE\\:(.*)"
```

Adding this file to a repository hooked into our CI system will create a single job that when executed run a
Windows batch command, and later on catches the message echoed and sets that as the build description.


# Tests #

This is an example of a simple application with tests:

```yaml

build_batch_commands:
- "pytest --junitxml=pytest_results.xml"

junit_patterns:
- "pytest_results.xml"
```

This jobs runs pytest in the repository and outputs test results to a file. We also configure the job to look for that
file, and present test results to us at then end of the build.

# Multiple platforms #


The same application as above, but now running on multiple platforms.

```yaml

platform-win64:build_batch_commands:
- "pytest --junitxml=pytest_results-{platform}.xml"

platform-redhat64:build_shell_commands:
- "pytest --junitxml=pytest_results-{platform}.xml"

junit_patterns:
- "pytest_results.*.xml"

matrix:
  platform:
  - "win64"
  - "redhat64"
```

Here we add a **matrix** section to define variations of this job. In this case, we have the platform variable,
with two possible values, `win64` and `redhat64`.

One job will be created for each possible combination in the matrix (only two jobs in this case).

Since we can't run batch commands in linux, we add another builder section, `build_shell_commands`. Using some flags
before defining sections we can choose which one will be available in each job.

Values from the matrix can also be used as variables, in this case, `{platform}` will be replaced by the platform used
in that job (`win64` or `redhat64`).


## Branch patterns ##

Branch patterns are used to filter which branches will produce jobs. This list of regular expressions
(using Python syntax), ensures that a branch will only produce jobs if at least one of the regular
expressions matches the name of the branch.

Here's an example that filter only `master` and feature branches:

```yaml

branch_patterns:
- "master"
- "fb-*"
```

If this section is not defined in the file, all branches will produce jobs.

## Job Matrix ##

As shown in the examples above, the job matrix can be used to create multiple variations of a job.
One job for each combination of entries in this matrix is created.

```yaml
matrix:
  mode:
  - "app"
  - "cases"
  platform:
  - "win64"
  - "linux64"
```

In this case 4 jobs will be generated:

* `app-win64`
* `app-linux64`
* `cases-win64`
* `cases-linux64`

Note that you can use any variables you need, Job's done has no idea what `mode` or `platform` means.

There's an `exclude` clause which can be used to remove particular entries from the matrix:


```yaml
matrix:
  mode:
  - "app"
  - "cases"
  platform:
  - "win64"
  - "linux64"

mode-cases:platform-win.*:exclude: "yes"
```

This will exclude all cases jobs from windows.

## String replacement ##

Variables defined in the job matrix can be used to replace strings in the job file.

On top of matrix variables, there are a few special string templates that can be used in any job:

* `name` - Name of the repository
* `branch` - Branch being built


```yaml
matrix:
  platform:
  - "win64"
  - "linux64"

platform-win.*:build_batch_commands:
- "echo Building project {name} in branch {branch} on platform {platform}"
```

Note that we use Python's format syntax, so if you need an actual `{` or `}` use double braces: `{{`, `}}`.


## Condition Flags ##

Variables defined in your job matrix can also be used to control some of the contents in your job file.
A common example here is using different builder for windows (bash) and linux (shell).

This is done by adding a prefix to sections of the YAML file, with the variable name and value necessary to use it:

```yaml
platform-win.*:build_batch_commands:
- "dir ."

platform-linux.*:build_shell_commands:
- "ls -la ."

matrix:
  platform:
  - "win32"
  - "linux64"
```

Matrix variables can also define aliases, useful to reduce duplication when using such flags.
To add aliases, simply use commas to separate additional names for matrix values:

```yaml
platform-windows:build_batch_commands:
- "dir ."

platform-linux:build_shell_commands:
- "ls -la ."

matrix:
  platform:
  - "win32,windows"
  - "win64,windows"
  - "linux64,linux"
```

On top of that you can use a special variable `branch` that's always available, and points to your branch:

```yaml
branch-master:build_batch_commands:
- "echo Build"

branch-deploy:build_batch_commands:
- "echo Build + Deploy"
```

Condition values can use Python regex syntax for extra flexibility:

```yaml
branch-master:build_batch_commands:
- "echo Build"

branch-fb.*:build_batch_commands:
- "echo Feature branch!"

branch-rb.*:build_batch_commands:
- "echo Release branch!"
```

# Development #

Create a virtual environment and install it in development mode:

```console
$ python -m virtualenv .env36
$ source .env36/bin/activate
$ pip install -e .
```

Run tests:

```console
$ pytest src
```

## Deploy ##

Open a PR updating the CHANGELOG and after it passes, push a tag to the repository.

This will automatically publish it to PyPI and to the configured Docker Registry in the organization.


# All options #

### additional_repositories ###

Additional repositories to be checked out in this job.

The repository where this .jobs_done file is included by default.

Requires [Multiple SCMs Plugin](https://wiki.jenkins-ci.org/display/JENKINS/Multiple+SCMs+Plugin) and [Git Plugin](https://wiki.jenkins-ci.org/display/JENKINS/Git+Plugin).

Uses same options as `git`.

```yaml
additional_repositories:
- git:
    url: "https://project.git"
    branch: "{branch}"
```

### auth_token ###

Job authentication token required to triggers builds remotely.

```yaml
auth_token: "my_token"
```

### boosttest_patterns ###

List of boosttest file patterns to look for test results.

Requires the [xUnit Plugin](https://wiki.jenkins-ci.org/display/JENKINS/xUnit+Plugin).

```yaml
boosttest_patterns:
- "*.xml"
```


### branch_patterns ###

List of regexes used to match branch names.
Only branches that match one of these will create
jobs.

```yaml
branch_patterns:
- "master"
- "fb-*"
```

### build_batch_commands ###
List of Windows batch commands used to build the job.
If errorcode is not 0 after any command, the build fails.

```yaml
build_batch_commands:
- "pytest src"
- "echo Finished"
```

### build_shell_commands ###

List of shell commands used to build the job.
If errorcode is not 0 after any command, the build fails.

```yaml
build_shell_commands:
- "pytest src"
- "echo Finished"
```

### build_python_commands ###

List of python commands used to build the job.

Requires [Python Plugin](https://wiki.jenkins-ci.org/display/JENKINS/Python+Plugin).

```yaml
build_python_commands:
- "print(5)"
```

### console_color ###

Enable support for ANSI escape sequences, including color, to Console Output.

Requires [AnsiColor Plugin](https://wiki.jenkins-ci.org/display/JENKINS/AnsiColor+Plugin).

Accepted values:
* `xterm` (default)
* `vga`
* `css`
* `gnome-terminal`

```yaml
console_color: "css"
```

### coverage ###


Enables code coverage report.

Require [Cobertura Plugin](https://wiki.jenkins-ci.org/display/JENKINS/Cobertura+Plugin).

Options:

* `report_pattern`: mandatory, pattern where XML coverage files are searched. These XML files are usually in
  [Cobertura](http://cobertura.github.io/cobertura) format, which is also format by
  [pytest-cov](https://pypi.python.org/pypi/pytest-cov) XML output (because pytest-cov uses coverage library).
* `healthy`: optional, specifies desired method, line and conditional metric. Any omitted metric defaults to `80`.
* `unhealthy`: optional, specifies desired method, line and conditional metric. Any omitted metric defaults to `0`. Builds below these thresholds are marked as unhealthy.
* `failing`: optional, specifies desired method, line and conditional metric. Any omitted metric defaults to `0`. Builds below these thresholds are marked as failed.

```yaml
coverage:
  report_pattern: "**/build/coverage/*.xml"
  healthy:
    method: 100
    line: 100
    conditional: 90
  unhealthy:
    method: 95
    line: 95
    conditional: 85
  failing:
    method: 90
    line: 90
    conditional: 80
```

### cron ###

Schedules to build to run periodically.

```yaml
cron: |
  # Everyday at 22pm
  * 22 * * *
```

### custom_workspace ###

Defines a custom workspace directory for the job. To maintain the same base directory as the default workspace directories prefix it with `"workspace/"`.

```yaml
custom_workspace: "workspace/devspace-user"
```

### description_regex ###

Regular expression for searching job output for a description.
If a match is found, the contents of the first group will be set as the description.

Requires [Description Setter Plugin](https://wiki.jenkins-ci.org/display/JENKINS/Description+Setter+Plugin).

```yaml
description_regex: "OUTPUT: (.*)"
```

### display_name ###

Configures the display name of the job.

```yaml
display_name: "{branch} {name}"
```

### email_notification ###

Sends emails for failed builds.

```yaml
email_notification: "email1@example.com email2@example.com"

# or

email_notification:
  recipients: "email1@example.com email2@example.com"
  notify_every_build: true
  notify_individuals: true
```

### exclude ###

Excludes a job from the matrix.

```yaml
platform-linux64:exclude: "yes"
```

### git ###

Additional git options for the main project.

Requires [Git Plugin](https://wiki.jenkins-ci.org/display/JENKINS/Git+Plugin).

Options available here are shared with `additional_repositories`.

```yaml
git:
  branch: master
  clean_checkout: false
  lfs: true
  recursive_submodules: true
  reference: "/home/reference.git"
  shallow_clone: 50
  tags: true
  target_dir: "main_application"
  timeout: 30
  url: "ssh://server/somerepo.git"
```

Note: by default, `tags` is `false` because we noticed that fetching tags takes quite some time and
we don't use tags for anything. If tags are needed, set `tags` to `true`.

### jsunit_patterns ###

List of jsunit file patterns to look for test results.

Requires [JSUnit Plugin](https://wiki.jenkins-ci.org/display/JENKINS/JSUnit+plugin).

```yaml
jsunit_patterns:
- "*.xml"
```

### junit_patterns ###

List of junit file patterns to look for test results.

Requires [xUnit Plugin](https://wiki.jenkins-ci.org/display/JENKINS/xUnit+Plugin).

```yaml
junit_patterns:
- "*.xml"
```

### label_expression ###

Configures the label expression of the job.

The label-expression is used to determine which workers can run the job.

```yaml
label_expression: "{platform}"
```

### matrix ###

Configures variations of a job.

```yaml
matrix:
  python:
  - "27"
  - "36"
```

### notify_github ###

Notifies GitHub of job build status.

Currently it does not support any parameters, using default values.

Requires [GitHub Plugin](https://plugins.jenkins.io/github).

```yaml
notify_github:
```

### notify_stash ###

Notifies a Stash instance of job build status.

When no parameters are given, uses configurations set in the Jenkins instance.

Requires [StashNotifier Plugin](https://wiki.jenkins-ci.org/display/JENKINS/StashNotifier+Plugin).

```yaml
notify_stash:
  url: "example.com/stash"
  username: "user"
  password: "pass"

# Using default from Jenkins
notify_stash:
```

### parameters ###

Job parameters for Jenkins jobs.

Currently, only `choice` and `string` are implemented.

```yaml
parameters:
  - choice:
      name: "PARAM_BIRD"
      choices:
        - "African"
        - "European"
      description: "Area where the bird is from"
  - string:
      name: "PARAM_VERSION"
      default: "dev"
      description: "App version"
```

### scm_poll ###

Schedules to periodically poll SCM for changes, and trigger builds.

```yaml
scm_poll: |
  # Everyday at 22pm
  * 22 * * *
```

### slack ###

Configure notification with slack.

1. Configure your Jenkins integration on Slack
2. Obtain the token
3. Configure your job to notify slack using this option.

```yaml
slack:
  team: esss
  channel: dev
  token: XXX
  url: https://example.com/jenkins
```

### timeout ###

Job timeout in minutes.

```yaml
timeout: 60
```

### timestamps ###

Show timestamps on the left side of the console output.

Requires the [Timestamper Plugin](https://wiki.jenkins-ci.org/display/JENKINS/Timestamper).

```yaml
timestamps:
```

### trigger_jobs ###

Trigger other jobs after the current job finishes. Parameters are optional.

```yaml
trigger_jobs:
  names:
    - myrepo-{branch}-synthetic-{platform}

CHANGELOG.md
# 1.11.0 (2022-07-28)

* The code has been updated to Python 3.10, including the Docker image.

# 1.10.2 (2022-07-27)

* Drop dependency to ``mailer``: this library no longer seems to be maintained, and is not installable in recent
  setuptools because it still uses the ``use_2to3`` option, which no longer exists in recent setuptools.

# 1.10.1 (2022-07-26)

* Fix error in Dockerfile: do not copy dummy `.env` file anymore, it needs to be supplied by the service running the container.

# 1.10.0 (2022-07-26)

* Changed the ``git/lfs`` option default to ``false``. Many hosting providers (GitHub included) charge LFS bandwidth
  usage, so it is common to use some kind of proxy to cache LFS files. Another reason is that ``false`` is the
  plugin default.

# 1.9.1 (2022-07-13)

* Fix bug where options for the ``git:`` setting were not being honored when set to ``false``.

# 1.9.0 (2022-07-08)

* Add support for GitHub build status notification with the `notify_github` option.

# 1.8.0 (2022-05-04)

* Split test dependencies into its own `dev` extra category (`pip install .[dev]` for example).
* Fixed handling of branch deletions and job updates in `/github`.
* The `GH_SECRET` has been renamed to `GH_WEBHOOK_SECRET` to make its purpose clearer.
* The `GH_USERNAME` setting is no longer used/needed.

# 1.7.1 (2022-04-27)

* Properly interpret the content type/mimedata of POST events.

# 1.7.0 (2022-04-26)

* The `/github` end-point now **requires** the webhook to be configured with a *secret*, which is known to JobsDone via the `JD_GH_SECRET` configuration variable.
* Tests have been moved to `tests`, outside the `src` directory. This avoids installing test-related files when using the package.

# 1.6.1 (2022-04-25)

* Add dependency to `gunicorn`, which was missing for deployment.

# 1.6.0 (2022-04-25)

* Server now understands GitHub pushes at the `/github` end-point.
* Server now understands Stash pushes also at the `/stash` end-point. Posting Stash pushes at `/` still works, but
  might be removed in the future.

# 1.5.0 (2021-11-11)

* New `tags` configuration added to Git, controlling if should fetch all tags or not during cloning. Default to `false`.

# 1.4.2 (2021-11-10)

* Fixed deploy to PyPI procedure.

# 1.4.0 (2021-10-14)

* Build commands (shell, batch, python) now flatten their lists allowing the use of references (#33).

# 1.3.1 (2021-09-10)

* Remove usage of xml.ElementTree.Element.getchildren for Python 3.9 compatibility.
* Use setuptools\_scm version for --version.

# 1.3.0 (2021-09-03)

* Revert parsable option `auto-updater`.
* Update generated XML for xUnit plugin to support most recent version.

# 1.2.2 (2019-09-11)

* Production version now logs to stdout.
* Refactor internal layout of Server app.

# 1.2.1 (2019-09-11)

* No code changes. Just build scripts

# 1.2.0 (2019-09-11)

* Add new parsable option `auto-updater`.

# 1.1.1 (2018-08-31)

* Drop unnecessary dependency to `jenkins-webapi`.

# 1.1.0 (2018-06-27)

- Now an empty JSON post will respond with 200 and the jobs_done10 version as plain text, instead of an error.

# 1.0.4 (2018-06-26)

- First release.


src/jobs_done10/server/app.py
# mypy: disallow-untyped-defs
import hmac
import json
import os
import pprint
import smtplib
import ssl
import traceback
from base64 import b64decode
from collections.abc import Callable
from collections.abc import Iterable
from collections.abc import Iterator
from email.mime.multipart import MIMEMultipart
from email.mime.text import MIMEText
from http import HTTPStatus
from typing import Any

import attr
import flask
import requests
from dotenv import load_dotenv
from pygments import highlight
from pygments.formatters import HtmlFormatter
from pygments.lexers import JsonLexer
from pygments.lexers import PythonTracebackLexer

from jobs_done10 import get_version_title
from jobs_done10.server_email_templates import EMAIL_HTML
from jobs_done10.server_email_templates import EMAIL_PLAINTEXT


app = flask.Flask("jobs_done")
load_dotenv(dotenv_path=os.environ.get("JOBSDONE_DOTENV"))
if not app.debug:
    app.logger.setLevel("INFO")
app.logger.info(f"Initializing Server App - {get_version_title()}")


class SignatureVerificationError(Exception):
    """
    Raised when we could not verify the authenticity of a post request.
    """


@app.route("/", methods=["GET", "POST"])
@app.route("/stash", methods=["GET", "POST"])
def stash() -> str | tuple[str, int]:
    """
    Jenkins job creation/update/deletion end-point for stash.
    """
    return _handle_end_point(parse_stash_post)


@app.route("/github", methods=["GET", "POST"])
def github() -> str | tuple[str, int]:
    """
    Jenkins job creation/update/deletion end-point for GitHub.
    """
    return _handle_end_point(parse_github_post)


def _handle_end_point(
    parse_request_callback: Callable[
        [dict[str, Any], dict[str, Any], bytes, dict[str, str]],
        Iterator["JobsDoneRequest"],
    ]
) -> str | tuple[str, int]:
    """Common handling for the jobs-done end-point."""
    request = flask.request
    payload = request.json
    json_payload_dump = (
        json.dumps(dict(payload), indent=2, sort_keys=True)
        if payload is not None
        else "None"
    )
    app.logger.info(
        "\n"
        + f"Received {request}\n"
        + f"Headers:\n"
        + json.dumps(dict(request.headers), indent=2, sort_keys=True)
        + "\n"
        + f"Payload (JSON):\n"
        + json_payload_dump
    )
    if request.method == "GET" or not request.data:
        # return a 200 response also on POST, when no data is posted; this is useful
        # because the "Test Connection" in BitBucket does just that, making it easy to verify
        # we have the correct version up.
        app.logger.info("I'm alive")
        return get_version_title()

    # Only accept JSON payloads.
    if payload is None:
        app.logger.info(f"POST body not in JSON format:\n{request.mimetype}")
        return (
            f"Only posts in JSON format accepted",
            HTTPStatus.BAD_REQUEST,
        )

    jobs_done_requests = []
    try:
        jobs_done_requests = list(
            parse_request_callback(
                request.headers, payload, request.data, dict(os.environ)
            )
        )
        app.logger.info(f"Parsed {len(jobs_done_requests)} jobs done requests:")
        for jdr in jobs_done_requests:
            app.logger.info("  - " + str(jdr) + "\n")
        message = _process_jobs_done_request(jobs_done_requests)
        app.logger.info(f"Output:\n{message}")
        return message
    except SignatureVerificationError as e:
        app.logger.exception(f"Header signature does not match: {e}")
        return str(e), HTTPStatus.FORBIDDEN
    except Exception:
        err_message = _process_jobs_done_error(
            request.headers, request.data, jobs_done_requests
        )
        app.logger.exception("Unexpected exception")
        return err_message, HTTPStatus.INTERNAL_SERVER_ERROR


@attr.s(auto_attribs=True, frozen=True)
class JobsDoneRequest:
    """
    Information necessary to process a jobs done request to create/update/delete jobs
    for a branch in a repository.
    """

    # Owner of the repository: "ESSS" for example.
    owner_name: str
    # Repository name: "etk" for example.
    repo_name: str
    pusher_email: str
    clone_url: str
    branch: str
    # A commit with None means a branch has been deleted.
    commit: str | None
    jobs_done_file_contents: str | None = attr.ib(repr=False)


def parse_stash_post(
    headers: dict[str, Any],
    payload: dict[str, Any],
    data: bytes,
    settings: dict[str, str],
) -> Iterator[JobsDoneRequest]:
    """
    Parses a Stash post information from a push event into jobs done requests.

    See ``_tests/test_server/stash-post.json`` for an example of a payload.
    """
    if not isinstance(payload, dict) or "eventKey" not in payload:
        raise RuntimeError(f"Invalid request json data: {pprint.pformat(payload)}")

    stash_url = settings["JD_STASH_URL"].rstrip()
    stash_username = settings["JD_STASH_USERNAME"]
    stash_password = settings["JD_STASH_PASSWORD"]
    project_key = payload["repository"]["project"]["key"]
    slug = payload["repository"]["slug"]
    pusher_email = payload["actor"]["emailAddress"]

    for change in payload["changes"]:
        branch = change["ref"]["id"]
        prefix = "refs/heads/"
        if not branch.startswith(prefix):
            continue

        branch = branch[len(prefix) :]

        jobs_done_file_contents = get_stash_file_contents(
            stash_url=stash_url,
            username=stash_username,
            password=stash_password,
            project_key=project_key,
            slug=slug,
            path=".jobs_done.yaml",
            ref=change["toHash"],
        )

        clone_url = get_stash_clone_url(
            stash_url=stash_url,
            username=stash_username,
            password=stash_password,
            project_key=project_key,
            slug=slug,
        )

        yield JobsDoneRequest(
            owner_name=project_key,
            repo_name=slug,
            clone_url=clone_url,
            commit=change["toHash"],
            pusher_email=pusher_email,
            branch=branch,
            jobs_done_file_contents=jobs_done_file_contents,
        )


def parse_github_post(
    headers: dict[str, Any],
    payload: dict[str, Any],
    data: bytes,
    settings: dict[str, str],
) -> Iterator[JobsDoneRequest]:
    """
    Parses a GitHub payload from a push event into jobs done requests.

    See ``_tests/test_server/github-post.json`` for an example of a payload.
    """
    verify_github_signature(headers, data, settings["JD_GH_WEBHOOK_SECRET"])
    owner_name = payload["repository"]["owner"]["login"]
    repo_name = payload["repository"]["name"]
    clone_url = payload["repository"]["ssh_url"]
    head_commit = payload["head_commit"]
    commit: str | None
    if head_commit is not None:
        commit = head_commit["id"]
    else:
        commit = None
    pusher_email = payload["pusher"]["email"]
    branch = payload["ref"]
    prefix = "refs/heads/"
    if branch.startswith(prefix):
        branch = branch[len(prefix) :]

    if commit is None:
        jobs_done_file_contents = None
    else:
        token = settings["JD_GH_TOKEN"]
        url = f"https://api.github.com/repos/{owner_name}/{repo_name}/contents/.jobs_done.yaml"
        response = requests.get(url, auth=("", token), params={"ref": commit})
        if response.status_code == HTTPStatus.NOT_FOUND:
            jobs_done_file_contents = None
        else:
            response.raise_for_status()
            payload = response.json()
            content = payload["content"]
            if payload["encoding"] != "base64":
                raise RuntimeError(
                    f"Unknown encoding when getting .jobs_done.yaml file: {payload['encoding']}"
                )
            jobs_done_file_contents = b64decode(content).decode("UTF-8")

    yield JobsDoneRequest(
        owner_name=owner_name,
        repo_name=repo_name,
        clone_url=clone_url,
        commit=commit,
        branch=branch,
        pusher_email=pusher_email,
        jobs_done_file_contents=jobs_done_file_contents,
    )


def verify_github_signature(headers: dict[str, Any], data: bytes, secret: str) -> None:
    """
    Verify the post raw data and our shared secret validate against the signature in the header.

    https://docs.github.com/en/developers/webhooks-and-events/webhooks/webhook-events-and-payloads#delivery-headers
    """
    signature_name = "x-hub-signature-256"
    try:
        header_signature = headers[signature_name]
    except KeyError:
        raise SignatureVerificationError(f'Missing "{signature_name}" entry in headers')
    algorithm = hmac.new(secret.encode("UTF-8"), data, "sha256")
    hash = algorithm.digest().hex()
    computed_signature = f"sha256={hash}"
    if not hmac.compare_digest(header_signature, computed_signature):
        raise SignatureVerificationError(
            f"Computed signature does not match the one in the header"
        )


def _process_jobs_done_request(jobs_done_requests: Iterable[JobsDoneRequest]) -> str:
    """
    Generate/update a Jenkins jobs from parsed requests and return a message
    informing what has been done.
    """
    from jobs_done10.generators import jenkins
    from jobs_done10.repository import Repository

    all_new_jobs = []
    all_updated_jobs = []
    all_deleted_jobs = []

    for request in jobs_done_requests:
        repository = Repository(url=request.clone_url, branch=request.branch)
        jenkins_url = os.environ["JD_JENKINS_URL"].rstrip("/")
        jenkins_username = os.environ["JD_JENKINS_USERNAME"]
        jenkins_password = os.environ["JD_JENKINS_PASSWORD"]

        new_jobs, updated_jobs, deleted_jobs = jenkins.UploadJobsFromFile(
            repository=repository,
            jobs_done_file_contents=request.jobs_done_file_contents,
            url=jenkins_url,
            username=jenkins_username,
            password=jenkins_password,
        )
        all_new_jobs.extend(new_jobs)
        all_updated_jobs.extend(updated_jobs)
        all_deleted_jobs.extend(deleted_jobs)

    lines: list[str] = []
    lines.extend(f"NEW - {x}" for x in all_new_jobs)
    lines.extend(f"UPD - {x}" for x in all_updated_jobs)
    lines.extend(f"DEL - {x}" for x in all_deleted_jobs)

    message = "\n".join(lines)
    return message


def _process_jobs_done_error(
    headers: dict[str, Any], data: bytes, jobs_done_requests: list[JobsDoneRequest]
) -> str:
    """
    In case of error while processing the job generation request, send an e-mail to the user with
    the traceback.
    """
    error_traceback = traceback.format_exc()
    payload = json.loads(data)
    lines = [
        f"ERROR processing request: {flask.request}",
        "",
        "Headers:",
        "",
        json.dumps(dict(headers), indent=2, sort_keys=True),
        "JSON data:",
        "",
        json.dumps(payload, indent=2, sort_keys=True),
        "",
        "",
        error_traceback,
        "",
    ]
    try:
        recipient = send_email_with_error(jobs_done_requests, payload, error_traceback)
    except Exception:
        lines.append("*" * 80)
        lines.append("ERROR SENDING EMAIL:")
        lines.append(traceback.format_exc())
    else:
        lines.append(f"Email sent to {recipient}")
    message = "\n".join(lines)
    return message


def get_stash_file_contents(
    *,
    stash_url: str,
    username: str,
    password: str,
    project_key: str,
    slug: str,
    path: str,
    ref: str,
) -> str | None:
    """
    Get the file contents from the stash server.

    We are using a "raw" Get which returns the entire file contents as text.
    """
    file_url = stash_url + f"/projects/{project_key}/repos/{slug}/raw/{path}?at={ref}"
    response = requests.get(file_url, auth=(username, password))
    if response.status_code == HTTPStatus.NOT_FOUND:
        return None
    response.raise_for_status()
    contents = response.text
    return contents


def get_stash_clone_url(
    *, stash_url: str, username: str, password: str, project_key: str, slug: str
) -> str:
    """
    Get information about the repository, returning the SSH clone url.

    It works by getting a request from the Stash server, see
    ``_tests/test_server/stash-repo-info.json`` for an example of a payload.
    """
    url = f"{stash_url}/rest/api/1.0/projects/{project_key}/repos/{slug}"
    response = requests.get(url, auth=(username, password))
    if response.status_code != 200:
        response.raise_for_status()

    data = response.json()
    for clone_url in data["links"]["clone"]:
        if clone_url["name"] == "ssh":
            return clone_url["href"]

    import pprint

    raise RuntimeError(
        f"Could not find the ssh clone url in json response:\n{pprint.pformat(data)}"
    )


def send_email_with_error(
    jobs_done_requests: list[JobsDoneRequest],
    payload: dict[str, Any],
    error_traceback: str,
) -> str:
    """
    Email the user who committed the changes that an error has happened while processing
    their .jobs_done file.

    Returns the recipient of the email in case of success, otherwise will raise an exception (not sure
    which exceptions are raised by the underlying library).
    """
    if not jobs_done_requests:
        raise RuntimeError("Cannot send an email without any parsed requests")

    pusher_email = jobs_done_requests[-1].pusher_email

    owner_name = jobs_done_requests[-1].owner_name
    repo_name = jobs_done_requests[-1].repo_name

    def abbrev_commit(c: str | None) -> str:
        return c[:7] if c is not None else "(no commit)"

    changes_msg = ", ".join(
        f"{x.branch} @ {abbrev_commit(x.commit)}" for x in jobs_done_requests
    )
    subject = (
        f"JobsDone failure during push to {owner_name}/{repo_name} ({changes_msg})"
    )

    pretty_json = pprint.pformat(payload)
    plain = EMAIL_PLAINTEXT.format(
        error_traceback=error_traceback, pretty_json=pretty_json
    )
    style = "colorful"
    html = EMAIL_HTML.format(
        error_traceback_html=highlight(
            error_traceback, PythonTracebackLexer(), HtmlFormatter(style=style)
        ),
        pretty_json_html=highlight(
            pretty_json, JsonLexer(), HtmlFormatter(style=style)
        ),
    )

    msg = MIMEMultipart()
    msg["Subject"] = subject
    msg["From"] = os.environ["JD_EMAIL_USER"]
    msg["To"] = pusher_email
    msg.attach(MIMEText(plain, "plain"))
    msg.attach(MIMEText(html, "html"))

    context = ssl.create_default_context()
    with smtplib.SMTP(
        os.environ["JD_EMAIL_SERVER"], int(os.environ["JD_EMAIL_PORT"])
    ) as server:
        server.starttls(context=context)
        server.login(os.environ["JD_EMAIL_USER"], os.environ["JD_EMAIL_PASSWORD"])
        server.sendmail(os.environ["JD_EMAIL_USER"], pusher_email, msg=msg.as_string())
    return pusher_email
